text/0118-priming-hints.md

```diff
@@ -0,0 +1,175 @@
+
+## Design
+
+We propose adding new meaningful comments to LWC which allow components to opt-in to marking properties that a priming mechanisms could assume to be true. This will allow priming mechanisms to prime a greater scope of component contents.
+
+These comments would be used on property getters, and their format would be:
+
+```js
+/**
+ * priming-hint: <scope> <value>
+ */
+```
+
+With the following values for "scope":
+- always: priming mechanisms should always evaluate the given property with the given value for the purposes of priming through the component tree
+- offline: priming mechanisms should evaluate the given property with the given value for the purposes of priming through the component tree only if the priming is targeted at an offline use case
+
+We differentiate between these 2 scopes because online and offline priming use cases can be qualitatively different. Online priming is aimed at enhancing performance of rendering, with the assumption that further resources and data can be pulled from the server later on as necessary. Offline priming aims to prime a complete useful set of resources and data from the server up-front with no additional requests later. In general, online priming is more sensitive to performance and offline priming is more sensitive to missing scope.
+
+For our bar example, we could modify the barItem component by replacing the use of a `@track` member with a new `@api` property that uses could make use of this priming hint to guide a priming mechanism to always prime its contents:
+
+```js
+export default class BarItem extends LightningElement {
+
+    /**
+     * priming-hint: always true
+     */
+    @api
+    viewContent() {
+        return this._loadContent;
+    }
+
+    @api
+    loadContent() {
+        this._loadContent = true;
+
+        this.dispatchEvent(new CustomEvent('active'));
+    }
+
+    ...
+}
+```
+
+```html
+<!-- barItem.html -->
+<template>
+    <template if:true={viewContent}>
+        <slot></slot>
+    </template>
+</template>
+```
+
+## Other Possible Strategies
+
+Here is a brief survey of other strategies we considered for priming hints.
+
+### Contextual Import
+
+```js
+import {primingContext} from `@salesforce/priming`;
+...
+get loadContent {
+    return primingContext?.isPriming || _loadContent;
+}
+```
+
+Here we have the property itself explicitly import an indicator of priming status and use it to drive the value to return for the property. This feels like it would be too much work for developers and they could easily calculate the property incorrectly.
+
+### Annotation
+
+```js
+import {forcePriming} from `@salesforce/priming`;
+...
+
+@forcePriming
+get loadContent {
+    return _loadContent;
+}
+```
+
+Here the component opts in to the priming-specific override by using a decorator. Using decorators like this seems out-of-step with other parts of the LWC specification.
```

Comment:
> @shesmer Why out of step? it could also be a decorator that allows more information like
@primingHint('always true')

I believe there were some concerns using decorators raised by either @kevinv11n or @pmdartus; unfortunately I don't have context on that conversation.

Reply:
The decorator proposal changed a lot since LWC adopted the syntax. The proposal was at stage 3 back in 2018, and then got retrograded to stage 2 because of implementors' concerns. It has been stuck at stage 2 for the last 3 years. Before adding new decorators, we want to ensure this proposal will actually ship. 

The good news is that the revisited proposal will be presented at the next TC39 meeting (in March) requesting stage 3.